lib/formatters/tapFormatter.mjs

```diff
@@ -57,3 +58,22 @@ export default function tapFormatter(results) {
 
 	return lines.join('\n');
 }
+
+/**
+ * replace octothorp/hash/hex/number sign/pound
+ * with its escape sequence
+ * in the file path or unique ID of the result's source
+ * @param {LintResult} result
+ * @modifies {result}
+ * @see {@link https://testanything.org/tap-version-14-specification.html#escaping TAP 14 specification}
+ */
+function escapeOctothorp(result) {
+	if (!result.source) return;
+
+	result.source = result.source.replaceAll('#', '%23');
```

Comment:
note

We don't use `\#` because:

1. postcss doesn't use a [custom alphabet](https://github.com/ai/nanoid#custom-alphabet-or-size) for the id
https://github.com/postcss/postcss/blob/main/lib/input.js#L61
i.e. `#` shouldn't be part of the id generated
2. in consequence, `#` may only be part of a file path
i.e. `encodeURIComponent('#') === '%23'`